Account/FirefoxAccount.swift

```diff
@@ -197,67 +198,52 @@ open class FirefoxAccount {
+                guard let url = url else {
+                    self.image = UIImage(named: "placeholder-avatar")
+                    return
+                }
+                
+                SDWebImageManager.shared().downloadImage(with: url, options: SDWebImageOptions.continueInBackground, progress: nil) { (image, error, cacheType, success, url) in
+                    guard image != nil else {
+                        self.downloadSucceded = false
+                        return
+                    }
+                    self.image = image
+                    NotificationCenter.default.post(name: NotificationFirefoxAccountProfileChanged, object: self)
+                }
             }
         }
     }
     
-    // Fetch current user's FxA profile. It contains the most updated email, displayName and avatar.
+    // Fetch current user's FxA profile. It contains the most updated email, displayName and avatar. This
+    // emits a `NotificationFirefoxAccountProfileChanged` once the profile has finished updating.
     open func updateProfile() {
         guard let session = stateCache.value as? TokenState else {
             return
         }
         
         let client = FxAClient10(authEndpoint: self.configuration.authEndpointURL, oauthEndpoint: self.configuration.oauthEndpointURL, profileEndpoint: self.configuration.profileEndpointURL)
-        return client.getProfile(withSessionToken: session.sessionToken as NSData) >>== { result in
-            
-            if let fxaProfile = self.fxaProfile {
-                // In the case there is an existing FxA profile, only update the changed properties.
-                if fxaProfile.email != result.email {
-                    self.fxaProfile = FxAProfile(email: result.email, displayName: result.displayName, avatar: result.avatar)
-                    self.fxaProfile?.downloadAvatarImage()
-                }
-                
-                fxaProfile.displayName = result.displayName
-                
-                if let avatar = result.avatar, let avatarUrl = URL(string: avatar) {
-                    if avatarUrl != fxaProfile.avatarUrl {
-                        fxaProfile.avatarUrl = avatarUrl
-                        fxaProfile.downloadAvatarImage()
-                    }
-                } else {
-                    fxaProfile.avatarUrl = nil
-                    fxaProfile.avatarImageData = nil
-                }
-            } else {
-                self.fxaProfile = FxAProfile(email: result.email, displayName: result.displayName, avatar: result.avatar)
-                self.fxaProfile?.downloadAvatarImage()
-            }
-            
+        client.getProfile(withSessionToken: session.sessionToken as NSData) >>== { result in
+            self.fxaProfile = FxAProfile(email: result.email, displayName: result.displayName, avatar: result.avatarURL)
             NotificationCenter.default.post(name: NotificationFirefoxAccountProfileChanged, object: self.fxaProfile)
```

Comment:
If the image is in the cache, this notification might fire very close in time, which might cause racing observers. Should we worry about this?

Reply:
I don't believe this is something to be worried about, but added a 100ms delay to posting notification after image has been downloaded.

Reply:
I don't think adding a short timed delay is a good fix, if there is a problem, and if there's no problem it adds unnecessary delay to UI updates. Please just inspect the observer notification handlers to make sure they're ready to receive notifications close in time and from arbitrary threads, and fix them if they're not.